File: ecommerce-app-ui/src/service/productService.ts
import type { Product, ProductsResponse } from "../store/types"
import api from "./axios"

export interface ProductQueryParams {
    category?:number;
    filter?:string;
    sort?:string;
    limit?:number;
    offset?:number;
}

export const getProducts = async (params: ProductQueryParams): Promise<ProductsResponse> => {
  const response = await api.get<ProductsResponse>('/products', { params })
  return response.data
}

export const getProductById = async (productId: number): Promise<Product> => {
  const response = await api.get<Product>(`/products/${productId}`)
  return response.data
}

File: e-commerce-backend/src/main/java/com/ecommerce/e_commerce_backend/service/ProductService.java
package com.ecommerce.e_commerce_backend.service;

import com.ecommerce.e_commerce_backend.dto.ProductDto;
import com.ecommerce.e_commerce_backend.dto.ProductsResponse;

public interface ProductService {
    ProductsResponse getProducts(
            Integer category,
            String filter,
            String sort,
            Integer limit,
            Integer offset
    );

    ProductDto getProductById(Long id);
}
